kratos/solving_strategies/convergencecriterias/residual_criteria.h

```diff
@@ -134,104 +125,114 @@ class ResidualCriteria : public  ConvergenceCriteria< TSparseSpace, TDenseSpace
+     * @param A System matrix (unused)
+     * @param Dx Vector of results (variations on nodal variables)
+     * @param b RHS vector (residual + reactions)
+     * @return true if convergence is achieved, false otherwise
+     */
     bool PostCriteria(
         ModelPart& rModelPart,
         DofsArrayType& rDofSet,
         const TSystemMatrixType& A,
         const TSystemVectorType& Dx,
         const TSystemVectorType& b
-    ) override
+        ) override
     {
-        if (TSparseSpace::Size(b) != 0) //if we are solving for something
-        {
+        const SizeType size_b = b.size();
+        if (size_b != 0) { //if we are solving for something
 
-            if (mInitialResidualIsSet == false)
-            {
-                mInitialResidualNorm = TSparseSpace::TwoNorm(b);
+            SizeType size_residual;
+            if (mInitialResidualIsSet == false) {
+//                 mInitialResidualNorm = TSparseSpace::TwoNorm(b); // NOTE: This doesn't take into account the reaction dofs
+                CalculateResidualNorm(mInitialResidualNorm, size_residual, rDofSet, b);
                 mInitialResidualIsSet = true;
 
-                //KRATOS_INFO(" Initial Residual ") << mInitialResidualNorm <<std::endl;
+                //KRATOS_INFO("RESIDUAL CRITERION") << "Initial Residual: " << mInitialResidualNorm <<std::endl;
             }
 
-            TDataType ratio;
-            mCurrentResidualNorm = TSparseSpace::TwoNorm(b);
+            TDataType ratio = 0.0;
+//             mCurrentResidualNorm = TSparseSpace::TwoNorm(b); // NOTE: This doesn't take into account the reaction dofs
+            CalculateResidualNorm(mCurrentResidualNorm, size_residual, rDofSet, b);
 
-            if(mInitialResidualNorm == 0.00)
-            {
-                ratio = 0.00;
-            }
-
-            else
-            {
+            if(mInitialResidualNorm < std::numeric_limits<TDataType>::epsilon()) {
+                ratio = 0.0;
+            } else {
                 ratio = mCurrentResidualNorm/mInitialResidualNorm;
             }
 
-            //KRATOS_INFO(" Current Residual ") << mCurrentResidualNorm << " ratio: "<< ratio << std::endl;
+            //KRATOS_INFO("RESIDUAL CRITERION") << "Current Residual: " <<  mCurrentResidualNorm << " ratio: "<< ratio << std::endl;
             
-	    double b_size = TSparseSpace::Size(b);
-	    TDataType absolute_norm = (mCurrentResidualNorm/b_size);
-			
-            if (rModelPart.GetCommunicator().MyPID() == 0)
-            {
-                if (this->GetEchoLevel() >= 1)
-                {
-                    std::cout << "RESIDUAL CRITERION :: Ratio = "<< ratio  << ";  Norm = " << absolute_norm << std::endl;
-                }
-            }
+            const TDataType float_size_residual = static_cast<TDataType>(size_residual);
+//             const TDataType float_size_residual = static_cast<TDataType>(size_b);
```

Comment:
at this point you can cleanup the commented lines

Reply:
Ok